[docs] Revise "Component theming" and "How to customize" guides (#31997)
mui/material-ui: docs/data/material/customization/theme-components/theme-components.md

# Components

<p class="description">Customizing the components using its keys inside the theme enables you to achieve styling consistency across your application.</p>

## Default props

Every Material UI component comes, out-of-the-box, with default values set for each of its props.

> If you want to change how your component looks by default, use the `defaultProps` key exposed in the theme's `components` key. Like so:

Reviewer: Trimming.

```suggestion
To change these default values, use the `defaultProps` key exposed in the theme's `components` key:
```

## Global style overrides

> If you want to print an entirely different design direction to the components, potentially changing every single style injected by Material UI into the DOM, use the theme's `styleOverrides` key.

Reviewer: I think these ideas should get their own sentences.

```suggestion
The theme's `styleOverrides` key makes it possible to potentially change every single style injected by Material UI into the DOM.
This is useful if you want to apply a fully custom design system to Material UI's components.
```

{{"demo": "GlobalThemeOverride.js"}}

> Note that each component is composed of several different parts. If you want to know how to target a specific part of a given component, check the **CSS** section of its API page to see every class it has available. You should then use it inside the `stylesOverrides` key.

Reviewer: Trying a different way of phrasing this for clarity.

```suggestion
Each component is composed of several different parts. 
These parts correspond to classes that are available to the component—see the **CSS** section of the component's API page for a detailed list.
You can use these classes inside the `styleOverrides` key to modify the corresponding parts of the component.
```

```js
const theme = createTheme({
  components: {
    MuiButton: {
      styleOverrides: {

>         containedPrimary: {

Reviewer 2: @siriwatknp as we are going in the direction of removing these, should we maybe change this demo a bit?

Reviewer 3: Good point, will do it.

> const theme = createTheme({
>   components: {
>     MuiButton: {
>       styleOverrides: {
>         containedPrimary: {
>           backgroundColor: '#202020',
>           color: '#fff',
>         },
>       },
>     },
>   },
> });

Reviewer 3: ```suggestion
const theme = createTheme({
  components: {
    MuiButton: {
      styleOverrides: ({ ownerState }) => ({
        ...ownerState.variant === 'contained' && ownerState.color === 'primary' && {
          backgroundColor: '#202020',
          color: '#fff',
        },
      }),
    },
  },
});
```

Author: I don't really have all developer context to judge whether this is better or not but I do need to confess that it just looks a bit uglier 😅

{{"demo": "GlobalThemeOverrideCallback.js"}}

### Using the `sx` experimental syntax

If you are not familiar with the `sx` prop, make sure you first check out [the concept around it](/system/the-sx-prop/) and [the difference with the `styled` utility](/system/styled/#difference-with-the-sx-prop).

To have a similar shorthand CSS notation to the `sx` prop but within the theme, you can use `sx` inside the `stylesOverrides`.

{{"demo": "GlobalThemeOverrideSx.js", "defaultCodeOpen": false}}

```tsx
const finalTheme = createTheme({
  components: {
    MuiChip: {
      styleOverrides: {
        root: sx({
          px: 1,
          py: 0.25,
          borderRadius: 1,
        }),
        label: {
          padding: 'initial',
        },
        icon: sx({
          mr: 0.5,
          ml: '-2px',
        }),
      },
    },
  },
});
```

### Specificity

If you use the theming approach to customize the components, you'll still be able to override them using the `sx` prop as it has a higher CSS specificity.

> Note that some components expose a `color` prop, which has a lower specificty than the theme's styles overrides. For example, say you have customized `Link`'s color.

Reviewer 3: @danilo-leal I was wrong. The `Link` is actually a bug. This should have the highest priority `<Link color="highest-priority-color">...</Link>`.

Author: Makes sense! I was just thinking it was weird for the component color prop to have lower specificity... I'll correct this sentence!

Reviewer 2: @siriwatknp are you going to bump the specificity for each prop combination? Should we create an RFC for this and make it work consistent with Material UI? In this case, when would you ever need the props callback?

Reviewer 3: @mnajdova I am not sure I follow you. From my understanding, this is a bug on the Link component.

```js

> MuiLink: {
>     styleOverrides: {
>       root: {
>        color: '#0000FF', // blue
>     },
>   },

Reviewer 3: ```suggestion
MuiLink: {
  styleOverrides: {
    root: {
       color: '#0000FF', // blue
    },
  },
},
```

{{"demo": "GlobalThemeOverrideCallback.js"}}

> ### Using the `sx` experimental syntax

Reviewer: Is the `sx` syntax only "experimental" in this specific context? I think we should elaborate on this if it's going to be mentioned in the header.

Author: I believe it is! Experimental in the sense that we're using it inside the theme object.

> If you are not familiar with the `sx` prop, make sure you first check out [the concept around it](/system/the-sx-prop/) and [the difference with the `styled` utility](/system/styled/#difference-with-the-sx-prop).

Reviewer: Trimming down & grammar.

```suggestion
If you are not familiar with the `sx` prop, learn more about [the concept behind it](/system/the-sx-prop/) and [how it differs from the `styled` utility](/system/styled/#difference-with-the-sx-prop).
```

Reviewer: On second thought, I think we should replace this sentence with a very brief explanation of the `sx` prop so the reader doesn't have to leave this page.

Author: Yup, I like that! I easily get exhausted when reading a piece of text that just several reference links... end up feeling like I have to go through all of them first before finishing the initial one I started with 😅 

How about that?

```suggestion
If you are not familiar with the `sx` prop, it consists of a shortcut for defining custom styles that access the theme object. The property is a superset of the CSS that packages all the style functions. To learn more about [the concept behind it](/system/the-sx-prop/) and [how it differs from the `styled` utility](/system/styled/#difference-with-the-sx-prop).
```

> To have a similar shorthand CSS notation to the `sx` prop but within the theme, you can use `sx` inside the `stylesOverrides`.

Reviewer: Rephrasing.

```suggestion
You can use the `sx` prop inside the `styleOverrides` key to modify styles within the theme using shorthand CSS notation.
```

Reviewer: Can we elaborate on why a developer would want to do this? And this would also be a good place to explain why this is "experimental."

Author: @mnajdova anything you'd like to share about that?

Reviewer 2: > Can we elaborate on why a developer would want to do this? 

Mainly for conistency. If a developer is already used to the `sx` syntax and it's shorthands, it's limiting that it can be used on only one place (the `sx` prop). More over, if they want to convert from `sx` prop to `styled()`, or move the styles to the theme overrides, they would have to manually change the styles to comply with the other regular syntax. This utility helps them by doing it automatically and providing a better DX along the way. 

It's experimental as it was added after the v5 stable release, and we usually export API like this as an experimental first, to see if there will be some unforeseen bugs and stabilize them later.

Author: Let me know what you think, tackled these aspects.

{{"demo": "GlobalThemeOverrideSx.js", "defaultCodeOpen": false}}

```tsx
const finalTheme = createTheme({
  components: {
    MuiChip: {
      styleOverrides: {
        root: sx({
          px: 1,
          py: 0.25,
          borderRadius: 1,
        }),
        label: {
          padding: 'initial',
        },
        icon: sx({
          mr: 0.5,
          ml: '-2px',
        }),
      },
    },
  },
});
```

### Specificity

> If you use the theming approach to customize the components, you'll still be able to override them using the `sx` prop as it has a higher CSS specificity.

Reviewer: This isn't entirely clear. Does this mean that `sx` in the theme overrides `sx` on individual components? Or vice versa?

Author: Oh, this is a great question... @mnajdova should know all the System answers 😃

Reviewer 3: `sx` > `theme variants` > `theme styleOverrides`.

{{"demo": "GlobalThemeOverrideCallback.js"}}

> ### Using the `sx` experimental syntax

Reviewer: This is nitpicking at this point but I think this formatting makes the most sense here.

```suggestion
### The `sx` syntax (experimental)
```

Author: I believe _syntax_ makes sense in this context as it is regarding its usage within the theme object, not as an actual component prop. 🤔

Reviewer 3: Add more info here. Technically, it is not a prop (React props) but a function. Maybe it is better to say "shorthand syntax" or "sx function"?

Author: I think syntax is accurate because it is pointing out how you can use `pr: 1` instead of `paddingRight: 1`.

> If you are not familiar with the `sx` prop, it consists of a shortcut for defining custom styles that access the theme object. The property is a superset of the CSS that packages all the style functions. To learn more about [the concept behind it](/system/the-sx-prop/) and [how it differs from the `styled` utility](/system/styled/#difference-with-the-sx-prop).

Reviewer: Trimming this down. 

```suggestion
The `sx` prop acts as a shortcut for defining custom styles that access the theme object. 
This prop lets you write inline styles using a superset of CSS. 
Learn more about [the concept behind the `sx` prop](/system/the-sx-prop/) and [how `sx` differs from the `styled` utility](/system/styled/#difference-with-the-sx-prop).
```

You can use the `sx` prop inside the `styleOverrides` key to modify styles within the theme using shorthand CSS notation.

> 

Author: @mnajdova do you have any insights?

> 

Author: Guess you already had in the above comment :) Will use it there.

{{"demo": "GlobalThemeOverrideCallback.js"}}

### The `sx` syntax (experimental)

The `sx` prop acts as a shortcut for defining custom styles that access the theme object.
This prop lets you write inline styles using a superset of CSS.
Learn more about [the concept behind the `sx` prop](/system/the-sx-prop/) and [how `sx` differs from the `styled` utility](/system/styled/#difference-with-the-sx-prop).

You can use the `sx` prop inside the `styleOverrides` key to modify styles within the theme using shorthand CSS notation.
This is especially handy if you're already using the `sx` prop with your components, because you can use the same syntax in your theme and quickly transfer styles between the two.

> > **Note:** The `sx` prop is a stable feature for customizing components in Material UI v5, but it is still considered _experimental_, since we released after the v5 stable launch, when used directly inside the theme object.

Reviewer: I think we can cut this clause.

```suggestion
> **Note:** The `sx` prop is a stable feature for customizing components in Material UI v5, but it is still considered _experimental_ when used directly inside the theme object.
```

This prop lets you write inline styles using a superset of CSS.
Learn more about [the concept behind the `sx` prop](/system/the-sx-prop/) and [how `sx` differs from the `styled` utility](/system/styled/#difference-with-the-sx-prop).

You can use the `sx` prop inside the `styleOverrides` key to modify styles within the theme using shorthand CSS notation.
This is especially handy if you're already used to using the `sx` prop syntax.

> It makes transferring one-off customizations using `sx` to the theme very fast.

Author: @samuelsycamore am I being too repetitive in mentioning `sx` three times in the same paragraph? 😅

Reviewer: I think it's always preferable to err on the side of accuracy if using pronouns would make things unclear.

If you wanted to rephrase it (totally optional), you could link the ideas in the last two sentences like this:

```suggestion
You can use the `sx` prop inside the `styleOverrides` key to modify styles within the theme using shorthand CSS notation.
This is especially handy if you're already using the `sx` prop with your components, because you can use the same syntax in your theme and quickly transfer styles between the two.
```
